fs-storage/src/file_storage.rs

```diff
@@ -0,0 +1,246 @@
+use std::fmt::Debug;
+use std::fs::{self, File};
+use std::io::{BufRead, BufReader, BufWriter, Write};
+use std::str::FromStr;
+use std::time::SystemTime;
+use std::{
+    collections::HashMap,
+    fmt::Display,
+    path::{Path, PathBuf},
+};
+
+use data_error::{ArklibError, Result};
+
+const LOG_PREFIX: &str = "[file-storage]";
+const STORAGE_VERSION: i32 = 2;
+const STORAGE_VERSION_PREFIX: &str = "version ";
+const KEY_VALUE_SEPARATOR: char = ':';
+
+pub struct FileStorage {
+    log_prefix: String,
+    label: String,
+    path: PathBuf,
+    timestamp: SystemTime,
+}
+
+impl FileStorage {
+    /// Create a new file storage with a diagnostic label and file path
+    pub fn new(label: String, path: &Path) -> Self {
+        Self {
+            label: label.clone(),
+            log_prefix: format!("{} {}", LOG_PREFIX, label).to_string(),
+            path: PathBuf::from(path),
+            timestamp: SystemTime::now(),
+        }
+    }
+
+    /// Read data from disk
+    ///
+    /// Data is read as a key value pairs separated by a symbol and stored
+    /// in a [HashMap] with a generic key K and V value. A handler
+    /// is called on the data after reading it.
+    pub fn read_file<K, V>(
+        &self,
+        mut handle: impl FnMut(HashMap<K, V>),
+    ) -> Result<()>
+    where
+        K: FromStr + std::hash::Hash + std::cmp::Eq + Debug,
+        V: FromStr + Debug,
+        ArklibError: From<<K as FromStr>::Err>,
+        ArklibError: From<<V as FromStr>::Err>,
+    {
+        let new_timestamp = fs::metadata(&self.path)?.modified()?;
+        log::info!(
+            "timestamp of storage file {:?} is {:?}",
+            self.path,
+            self.timestamp
+        );
+
+        if self.timestamp >= new_timestamp {
+            return Ok(());
+        }
+
+        log::info!("the file was modified externally, merging");
+
+        let value_by_id = self.read_file_from_disk()?;
+        if !value_by_id.is_empty() {
+            handle(value_by_id);
+        }
+
+        Ok(())
```

Comment:
It seems that `self.timestamp` is not updated as it's done in Kotlin implementation after each read:
https://github.com/ARK-Builders/arklib-android/blob/main/lib/src/main/java/dev/arkbuilders/arklib/data/storage/FileStorage.kt#L99

Reply:
Fixed the logic to update timestamp from metadata of read file.